File: final_assignment_front/lib/utils/workflow_permissions.dart
import 'package:final_assignment_front/utils/helpers/app_helpers.dart';

bool canApprove(String? status) {
  final state = AppealProcessStatus.fromCode(status);
  final normalized =
      status?.replaceAll('-', '_').replaceAll(' ', '_').toLowerCase();
  return state == AppealProcessStatus.underReview ||
      normalized == 'under_review' ||
      normalized == 'underreview';
}

bool canReject(String? status) => canApprove(status);

bool canStartReview(String? status) {
  final state = AppealProcessStatus.fromCode(status);
  if (state == AppealProcessStatus.unprocessed) {
    return true;
  }
  final normalized =
      status?.replaceAll('-', '_').replaceAll(' ', '_').toLowerCase();
  return normalized == 'pending' ||
      normalized == 'unprocessed' ||
      normalized == 'under_review_pending';
}

bool canEdit(String? status) {
  final appealState = AppealProcessStatus.fromCode(status);
  if (appealState != null) {
    return appealState != AppealProcessStatus.approved &&
        appealState != AppealProcessStatus.rejected &&
        appealState != AppealProcessStatus.withdrawn;
  }

  final offenseState = OffenseProcessStatus.fromCode(status);
  if (offenseState != null) {
    return offenseState == OffenseProcessStatus.unprocessed ||
        offenseState == OffenseProcessStatus.processing;
  }

  final paymentState = PaymentStatus.fromCode(status);
  if (paymentState != null) {
    return paymentState == PaymentStatus.unpaid ||
        paymentState == PaymentStatus.partial ||
        paymentState == PaymentStatus.overdue;
  }

  return false;
}

bool canPay(String? status) {
  final state = PaymentStatus.fromCode(status);
  return state == PaymentStatus.unpaid ||
      state == PaymentStatus.partial ||
      state == PaymentStatus.overdue;
}

File: final_assignment_front_react/src/utils/workflowPermissions.js
import { STATUSES } from '../constants/statuses.js';
import { STATUS } from './statusLabels.js';

export const APPEAL_PROCESS_STATUS = Object.freeze({
  unprocessed: STATUS.UNPROCESSED,
  underReview: STATUS.UNDER_REVIEW,
  approved: STATUSES.APPROVED,
  rejected: STATUSES.REJECTED,
  withdrawn: STATUS.WITHDRAWN,
});

export const APPEAL_PROCESS_EVENT = Object.freeze({
  approve: 'APPROVE',
  reject: 'REJECT',
});

const REVIEWABLE_APPEAL_STATUSES = new Set([
  APPEAL_PROCESS_STATUS.unprocessed,
  APPEAL_PROCESS_STATUS.underReview,
]);

const TERMINAL_APPEAL_STATUSES = new Set([
  APPEAL_PROCESS_STATUS.approved,
  APPEAL_PROCESS_STATUS.rejected,
  APPEAL_PROCESS_STATUS.withdrawn,
]);

export const canApprove = (status) => REVIEWABLE_APPEAL_STATUSES.has(status);

export const canReject = (status) => canApprove(status);

export const canEdit = (status) =>
  Boolean(status) && !TERMINAL_APPEAL_STATUSES.has(status);
